File: Python/OneLang/Template/TemplateParser.py
from onelang_core import *
import OneLang.Parsers.Common.Reader as read
import OneLang.Template.Nodes as nodes
import OneLang.One.Ast.Expressions as exprs
import OneLang.Parsers.TypeScriptParser as typeScrPars
import re

class BlockIndentManager:
    def __init__(self, block, indent_len):
        self.deindent_len = -1
        self.block = block
        self.indent_len = indent_len
    
    def remove_prev_indent(self):
        if len(self.block.items) == 0:
            return 0
        last_item = self.block.items[len(self.block.items) - 1]
        
        if not (isinstance(last_item, nodes.LiteralNode)):
            return 0
        lit = last_item
        
        pos = len(lit.value) - 1
        
        while pos >= 0:
            if lit.value[pos] == "\n":
                indent = len(lit.value) - pos - 1
                lit.value = lit.value[0:pos]
                if indent < 0:
                    pass
                return indent
            
            if lit.value[pos] != " ":
                break
            pos = pos - 1
        
        return 0
    
    def deindent(self, lit):
        if self.indent_len == 0:
            return lit
        # do not deindent root nodes
               
        lines = re.split("\\r?\\n", lit.value)
        if len(lines) == 1:
            return lit
        
        new_lines = [lines[0]]
        i_line = 1
        
        while i_line < len(lines):
            line = lines[i_line]
            
            if self.deindent_len == -1:
                i = 0
                
                while i < len(line):
                    if line[i] != " ":
                        self.deindent_len = i
                        if self.deindent_len - self.indent_len < 0:
                            pass
                        break
                    i = i + 1
            
            if self.deindent_len == -1:
                new_lines.append(line)
            else:
                space_len = len(line) if len(line) < self.deindent_len else self.deindent_len
                i = 0
                
                while i < space_len:
                    if line[i] != " ":
                        raise Error("invalid indent")
                    i = i + 1
                new_lines.append(line[self.deindent_len - self.indent_len:])
            i_line = i_line + 1
        lit.value = "\n".join(new_lines)
        return lit

class TemplateParser:
    def __init__(self, template):
        self.reader = None
        self.parser = None
        self.template = template
        self.parser = typeScrPars.TypeScriptParser2(template)
        self.parser.allow_dollar_ids = True
        self.reader = self.parser.reader
    
    def parse_attributes(self):
        result = {}
        while self.reader.read_token(","):
            key = self.reader.expect_identifier()
            value = self.reader.expect_string() if self.reader.read_token("=") else None
            result[key] = value
        return result
    
    def parse_block(self, indent_len = 0):
        block = nodes.TemplateBlock([])
        indent_man = BlockIndentManager(block, indent_len)
        while not self.reader.get_eof():
            if self.reader.peek_exactly("{{/"):
                break
            if self.reader.read_exactly("${"):
                expr = self.parser.parse_expression()
                block.items.append(nodes.ExpressionNode(expr))
                self.reader.expect_token("}")
            elif self.reader.read_exactly("$"):
                id = self.reader.expect_identifier()
                block.items.append(nodes.ExpressionNode(exprs.Identifier(id)))
            elif self.reader.read_exactly("{{"):
                block_indent_len = indent_man.remove_prev_indent()
                
                if self.reader.read_token("for"):
                    var_name = self.reader.expect_identifier()
                    self.reader.expect_token("of")
                    items_expr = self.parser.parse_expression()
                    attrs = self.parse_attributes()
                    self.reader.expect_token("}}")
                    body = self.parse_block(block_indent_len)
                    self.reader.expect_token("{{/for}}")
                    block.items.append(nodes.ForNode(var_name, items_expr, body, attrs.get("joiner")))
                else:
                    expr = self.parser.parse_expression()
                    block.items.append(nodes.ExpressionNode(expr))
                    self.reader.expect_token("}}")
            else:
                literal = self.reader.read_regex("([^\\\\]\\\\(\\{\\{|\\$)|\r|\n|(?!\\{\\{|\\$\\{|\\$).)*")[0]
                if literal == "":
                    raise Error("This should not happen!")
                block.items.append(indent_man.deindent(nodes.LiteralNode(literal)))
        
        if indent_len != 0:
            indent_man.remove_prev_indent()
        
        return block
    
    def parse(self):
        return self.parse_block()

File: PHP/src/OneLang/Template/TemplateParser.php
<?php

namespace OneLang\Template\TemplateParser;

use OneLang\Parsers\Common\Reader\Reader;
use OneLang\Template\Nodes\ExpressionNode;
use OneLang\Template\Nodes\ForNode;
use OneLang\Template\Nodes\ITemplateNode;
use OneLang\Template\Nodes\LiteralNode;
use OneLang\Template\Nodes\TemplateBlock;
use OneLang\One\Ast\Expressions\Identifier;
use OneLang\Parsers\TypeScriptParser\TypeScriptParser2;

class BlockIndentManager {
    public $deindentLen;
    public $block;
    public $indentLen;
    
    function __construct($block, $indentLen) {
        $this->block = $block;
        $this->indentLen = $indentLen;
        $this->deindentLen = -1;
    }
    
    function removePrevIndent() {
        if (count($this->block->items) === 0)
            return 0;
        $lastItem = $this->block->items[count($this->block->items) - 1];
        
        if (!($lastItem instanceof LiteralNode))
            return 0;
        $lit = $lastItem;
        
        for ($pos = strlen($lit->value) - 1; $pos >= 0; $pos--) {
            if ($lit->value[$pos] === "\n") {
                $indent = strlen($lit->value) - $pos - 1;
                $lit->value = substr($lit->value, 0, $pos - (0));
                if ($indent < 0) { }
                return $indent;
            }
            
            if ($lit->value[$pos] !== " ")
                break;
        }
        
        return 0;
    }
    
    function deindent($lit) {
        if ($this->indentLen === 0)
            return $lit;
        // do not deindent root nodes
               
        $lines = preg_split("/\\r?\\n/", $lit->value);
        if (count($lines) === 1)
            return $lit;
        
        $newLines = array($lines[0]);
        for ($iLine = 1; $iLine < count($lines); $iLine++) {
            $line = $lines[$iLine];
            
            if ($this->deindentLen === -1)
                for ($i = 0; $i < strlen($line); $i++) {
                    if ($line[$i] !== " ") {
                        $this->deindentLen = $i;
                        if ($this->deindentLen - $this->indentLen < 0) { }
                        break;
                    }
                }
            
            if ($this->deindentLen === -1)
                $newLines[] = $line;
            else {
                $spaceLen = strlen($line) < $this->deindentLen ? strlen($line) : $this->deindentLen;
                for ($i = 0; $i < $spaceLen; $i++) {
                    if ($line[$i] !== " ")
                        throw new \OneLang\Core\Error("invalid indent");
                }
                $newLines[] = substr($line, $this->deindentLen - $this->indentLen);
            }
        }
        $lit->value = implode("\n", $newLines);
        return $lit;
    }
}

class TemplateParser {
    public $reader;
    public $parser;
    public $template;
    
    function __construct($template) {
        $this->template = $template;
        $this->parser = new TypeScriptParser2($template);
        $this->parser->allowDollarIds = true;
        $this->reader = $this->parser->reader;
    }
    
    function parseAttributes() {
        $result = Array();
        while ($this->reader->readToken(",")) {
            $key = $this->reader->expectIdentifier();
            $value = $this->reader->readToken("=") ? $this->reader->expectString() : null;
            $result[$key] = $value;
        }
        return $result;
    }
    
    function parseBlock($indentLen = 0) {
        $block = new TemplateBlock(array());
        $indentMan = new BlockIndentManager($block, $indentLen);
        while (!$this->reader->get_eof()) {
            if ($this->reader->peekExactly("{{/"))
                break;
            if ($this->reader->readExactly("\${")) {
                $expr = $this->parser->parseExpression();
                $block->items[] = new ExpressionNode($expr);
                $this->reader->expectToken("}");
            }
            else if ($this->reader->readExactly("\$")) {
                $id = $this->reader->expectIdentifier();
                $block->items[] = new ExpressionNode(new Identifier($id));
            }
            else if ($this->reader->readExactly("{{")) {
                $blockIndentLen = $indentMan->removePrevIndent();
                
                if ($this->reader->readToken("for")) {
                    $varName = $this->reader->expectIdentifier();
                    $this->reader->expectToken("of");
                    $itemsExpr = $this->parser->parseExpression();
                    $attrs = $this->parseAttributes();
                    $this->reader->expectToken("}}");
                    $body = $this->parseBlock($blockIndentLen);
                    $this->reader->expectToken("{{/for}}");
                    $block->items[] = new ForNode($varName, $itemsExpr, $body, (@$attrs["joiner"] ?? null));
                }
                else {
                    $expr = $this->parser->parseExpression();
                    $block->items[] = new ExpressionNode($expr);
                    $this->reader->expectToken("}}");
                }
            }
            else {
                $literal = $this->reader->readRegex("([^\\\\]\\\\(\\{\\{|\\\$)|\r|\n|(?!\\{\\{|\\\$\\{|\\\$).)*")[0];
                if ($literal === "")
                    throw new \OneLang\Core\Error("This should not happen!");
                $block->items[] = $indentMan->deindent(new LiteralNode($literal));
            }
        }
        
        if ($indentLen !== 0)
            $indentMan->removePrevIndent();
        
        return $block;
    }
    
    function parse() {
        return $this->parseBlock();
    }
}
